infinite_looping_in_forward_transaction_#931

apoorv-2204 opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes 

- [ ] #931 

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- Test A
- Test B

# Checklist:

- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

apoorv-2204 commented on lib/archethic.ex:
```diff
@@ -74,25 +74,49 @@ defmodule Archethic do
+        nodes
+      end
+
+    spawn(fn ->
```
its lightweight and much better than tasks supervisor

samuelmanzanera replied:
You should not let process unsupervised in your system, as it can lead to unexpected issues and problems. Agree with @neylix

bchamagne commented on lib/archethic/p2p/message/new_transaction.ex:
```diff
@@ -17,13 +17,8 @@ defmodule Archethic.P2P.Message.NewTransaction do
-        %Error{reason: :network_issue}
-    end
+    :ok = Archethic.send_new_transaction(tx, node_pbkey)
+    %Ok{}
```
why did you remove all the error cases?

bchamagne commented on lib/archethic.ex:
```diff
@@ -74,25 +73,50 @@ defmodule Archethic do
+    nodes =
+      P2P.authorized_and_available_nodes()
+      |> Enum.reject(&(&1.first_public_key == welcome_node_key))
+      |> Enum.sort_by(& &1.first_public_key)
```
is this `sort_by` useful? look like the `nearest_nodes` is resorting it right after?

Neylix replied:
This sort is usefull when there is multiple nodes in the same patch. We need to have the same order of node within the same patch, while the nearest_nodes function does not ensure the sort inside a patch

bchamagne commented on lib/archethic.ex:
```diff
@@ -74,25 +73,50 @@ defmodule Archethic do
+        # next node from the this node position in nodes list
+        current_node = Crypto.first_node_public_key()
+
+        index = Enum.find_index(nodes, &(&1.first_public_key == current_node))
```
since there is a sort by nearest_nodes, is the order of the nodes determinist?

Neylix replied:
The sort is based on the network patch if the welcome node, not the local node one.
So as the network patch of the welcome is the same across all node, the sort result in the same ordre

samuelmanzanera replied:
Yes sure, I didn't see the welcome_node_patch, so definitively as long as the network remains the same across the nodes, it should be deterministic.

Neylix reviewed (commented):
Looks good to me, can your rebase to solve conflict and not override new changes ?

Neylix commented on test/archethic_test.exs:
```diff
@@ -130,6 +241,101 @@ defmodule ArchethicTest do
+      assert :ok = Archethic.send_new_transaction(tx)
+
+      # start repair and should bottleneck requests
+      pid = SelfRepair.repair_in_progress?(nss_genesis_address)
```
I unresolve the thread as this as not been changed

apoorv-2204 replied:
i deleted the test

Neylix commented on test/archethic_test.exs:
```diff
@@ -130,6 +241,101 @@ defmodule ArchethicTest do
+      assert pid != nil
+
+      assert_receive({:forwarded_to_node2, ^tx})
+      Process.sleep(100)
```
I unresolve the thread as this as not been changed

samuelmanzanera commented on lib/archethic.ex:
```diff
@@ -80,25 +79,50 @@ defmodule Archethic do
+      end
+
+    TaskSupervisor
+    |> Task.Supervisor.async_nolink(fn ->
```
Why do you need other process to send the transaction? 
And also, what is the purpose of the `async`? You are not awaiting the task result.

apoorv-2204 replied:
spawn?

apoorv-2204 replied:
node 1 waits for node 2 to return ok , node 2 waits for node 3 to send txn and fget back response,

samuelmanzanera commented on lib/archethic.ex:
```diff
@@ -80,25 +79,50 @@ defmodule Archethic do
+      |> do_forward_transaction(nodes)
+    end)
+
+    :ok
```
The network issue is never returned here? I think it would still be useful.

bchamagne reviewed (commented):
To make the test work I had to hardcode a network patch, because there is a randomness in local development that breaks this feature. 

I tested this scenario: 

:heavy_check_mark:  3 nodes, I modified the code of 2 nodes to mark them as invalid for processing (to force the forward).  I then sent a transaction to `A` that forward it to `B` that forward it to `C`. 

- I think your last modification modify the return type of `send_new_transaction`, is this ok?
- When you remove the task.async, we now have `A` that wait until timeout, and when the timeout occur, `A` will try to send to `C` (in the same scenario as above), which I believe is not good. 

ps: please keep commit history

apoorv-2204 commented:
> To make the test work I had to hardcode a network patch, because there is a randomness in local development that breaks this feature.
> 
> I tested this scenario:
> 
> ✔️ 3 nodes, I modified the code of 2 nodes to mark them as invalid for processing (to force the forward). I then sent a transaction to `A` that forward it to `B` that forward it to `C`.
> 
> * When you remove the task.async, we now have `A` that wait until timeout, and when the timeout occur, `A` will try to send to `C` (in the same scenario as above), which I believe is not good.
> 
> ps: please keep commit history

Hi, we are relying on timeout to notifiy user. I updated geo patch for null patch for localhost

samuelmanzanera commented:
> > To make the test work I had to hardcode a network patch, because there is a randomness in local development that breaks this feature.
> > I tested this scenario:
> > ✔️ 3 nodes, I modified the code of 2 nodes to mark them as invalid for processing (to force the forward). I then sent a transaction to `A` that forward it to `B` that forward it to `C`.
> > 
> > * When you remove the task.async, we now have `A` that wait until timeout, and when the timeout occur, `A` will try to send to `C` (in the same scenario as above), which I believe is not good.
> > 
> > ps: please keep commit history
> 
> Hi, we are relying on timeout to notifiy user. I updated geo patch for null patch for localhost

Indeed, I wasn't against the spawn a process it's just the Task.async expected to await something, maybe a start_child would be sufficient. (But QUID of the error network_issue then ?)
Otherwise we will have cascading timeout.

bchamagne commented:
Maybe I got it wrong, but I believe my last comment is still relevant. We talked about an ACK msg no?

apoorv-2204 commented:
> Maybe I got it wrong, but I believe my last comment is still relevant. We talked about an ACK msg no?

Replay back a message?

Neylix commented:
> Maybe I got it wrong, but I believe my last comment is still relevant. We talked about an ACK msg no?

The ack message is returned in the `process` function of the message `NewTransaction`
